server/src/service.rs

```diff
@@ -458,4 +467,50 @@ impl Service {
 
         Ok(signal_emitter)
     }
+
+    pub async fn session_from_client_name(&self, client_name: &str) -> Option<Session> {
+        let sessions = self.sessions.lock().await;
+
+        for session in sessions.values() {
+            if session.client_name() == client_name {
+                return Some(session.clone());
+            }
+        }
+
+        None
+    }
+
+    pub async fn on_client_disconnect(&self) -> Result<(), Error> {
+        let rule = zbus::MatchRule::builder()
+            .msg_type(zbus::message::Type::Signal)
+            .sender("org.freedesktop.DBus")?
+            .interface("org.freedesktop.DBus")?
+            .member("NameOwnerChanged")?
+            .arg(2, "")?
+            .build();
+        let mut stream = zbus::MessageStream::for_match_rule(rule, &self.connection, None).await?;
+
+        while let Some(message) = stream.try_next().await? {
+            let Ok((_name, old_owner, new_owner)) =
+                message.body().deserialize::<(String, String, String)>()
+            else {
+                continue;
+            };
+
```

Comment:
can this ever fail to deserialize? also add the assertion that new_owner is empty, because you have it in the match rule...